glean-core/ios/Glean/Net/HttpPingUploader.swift

```diff
@@ -53,11 +53,52 @@ public class HttpPingUploader {
     /// headers are added to the HTTP request in addition to the UserAgent. This allows
     /// us to easily handle pings coming from Glean on the legacy Mozilla pipeline.
     func upload(path: String, data: Data, headers: [String: String], callback: @escaping (UploadResult) -> Void) {
-        // Build the request and create an async upload operation and launch it through the
-        // Dispatchers
-        if let request = buildRequest(path: path, data: data, headers: headers) {
-            let uploadOperation = PingUploadOperation(request: request, data: data, callback: callback)
-            Dispatchers.shared.launchConcurrent(operation: uploadOperation)
+        // Build the request and create an async upload operation using a background URLSession
+        if let request = self.buildRequest(path: path, data: data, headers: headers) {
+            // Build a URLSession with no-caching suitable for uploading our pings
+            let config: URLSessionConfiguration
+            if Dispatchers.shared.testingMode {
+                // For test mode, we want the URLSession to send things ASAP, rather than in the background
+                config = URLSessionConfiguration.default
+            } else {
+                // For normal use cases, we will take advantage of the background URLSessionConfiguration
+                // which will pass the data to the OS as a file and the OS will then handle the request
+                // in a separate process
+                config = URLSessionConfiguration.background(withIdentifier: path)
+            }
+            config.sessionSendsLaunchEvents = false // We don't need to notify the app when we are done.
+            config.requestCachePolicy = NSURLRequest.CachePolicy.reloadIgnoringLocalCacheData
+            config.isDiscretionary = false
+            config.urlCache = nil
+            let session = URLSession(configuration: config,
+                                     delegate: SessionResponseDelegate(callback),
+                                     delegateQueue: Dispatchers.shared.serialOperationQueue)
+
+            let tmpFile = URL.init(fileURLWithPath: NSTemporaryDirectory(),
+                                   isDirectory: true).appendingPathComponent("\(path.split(separator: "/").last!)")
+            do {
+                try data.write(to: tmpFile, options: .noFileProtection)
+            } catch {
+                // Since we cannot write the file, there is no need to continue and schedule an
+                // upload task. So instead we log the error and return.
+                self.logger.error("\(error)")
+                return
```

Comment:
As per documentation the `URLSession` "keeps a strong reference to the delegate until your app exits or explicitly invalidates the session".
We should just move creating the `URLSession` until after this do/catch (or we need to call `session.invalidateAndCancel()` to invalidate it)